source/particles/particle_handler.cc

```diff
@@ -1346,17 +1346,20 @@ namespace Particles
           const unsigned int closest_vertex =
             GridTools::find_closest_vertex_of_cell<dim, spacedim>(
               current_cell, out_particle->get_location(), *mapping);
+
+          const unsigned int closest_vertex_index =
+            current_cell->vertex_index(closest_vertex);
+
           Tensor<1, spacedim> vertex_to_particle =
             out_particle->get_location() - current_cell->vertex(closest_vertex);
 
           // Only normalize if vector is larger than zero, the algorithm below
           // works fine for zero vectors.
-          if (vertex_to_particle.norm() >
-              100. * std::numeric_limits<double>::epsilon())
+          if (vertex_to_particle.norm_square() >
+              100. * std::numeric_limits<double>::epsilon() *
+                vertex_to_cell_centers[closest_vertex_index][0].norm_square())
             vertex_to_particle /= vertex_to_particle.norm();
```

Comment:
As explained in the other PR, I think you should add
```
  else
    vector_to_particle = 0;
```
and think about whether everything you call below still works in that case (and/or if it could be simplified). That's because if you decide that the `vertex_to_particle` vector is *essentially* zero, then it contains only numerical noise and nothing good can happen if you actually use that.

Reply:
Also, if you don't square epsilon, you're essentially setting a cut-off at 1e-7 (relative) here. That seems large to me.